cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,523 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package simulator
+
+import (
+	"fmt"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	schedulerlisters "k8s.io/kubernetes/pkg/scheduler/listers"
+	schedulernodeinfo "k8s.io/kubernetes/pkg/scheduler/nodeinfo"
+)
+
+// DeltaClusterSnapshot is implementation of ClusterSnapshot optimized for typical Cluster Autoscaler usage - (fork, add stuff, revert) repeated many times per loop.
+//
+// Complexity of some notable operations:
+//	fork - O(1)
+//	revert - O(1)
+//	commit - O(n)
+//	list all pods (no filtering) - O(n), cached
+//	list all pods (with filtering) - O(n)
+//	list node infos - O(n), cached
+//	get all nodes - O(n)
+//
+// Watch out for:
+//	node deletions, pod additions & deletions - invalidates cache of current snapshot
+//		(when forked affects delta, but not base.)
+//	pod affinity - causes scheduler framework to list pods with non-empty selector,
+//		so basic caching doesn't help.
+//
+type DeltaClusterSnapshot struct {
+	data *internalDeltaSnapshotData
+}
+
+type internalDeltaSnapshotDataNodeLister internalDeltaSnapshotData
+type internalDeltaSnapshotDataPodLister internalDeltaSnapshotData
+
+type internalDeltaSnapshotData struct {
+	baseData *internalDeltaSnapshotData
+
+	nodeInfoMap      map[string]*schedulernodeinfo.NodeInfo
+	deletedNodeInfos map[string]bool
+
+	nodeInfoList         []*schedulernodeinfo.NodeInfo
```

Comment:
Acutally do we need both `nodeInfoList` and `completeNodeInfoList`?
It look like you are only using first to build the other - and it feels like having just the second one would be enough.
Lets disucss this offline.

Reply:
Merged